doc/developer/design/20230921_distributed_ts_oracle.md

```diff
@@ -0,0 +1,442 @@
+    RETURNING write_ts;
+```
+
+`peek_write_ts(timeline)`:
+
+```sql
+SELECT write_ts FROM timestamp_oracle
+    WHERE timeline = $timeline;
+```
+
+`get_read_ts(timeline)`:
+
+```sql
+SELECT read_ts FROM timestamp_oracle
+    WHERE timeline = $timeline;
+```
+
+`apply_write(timeline, write_ts)`:
+
+```sql
+UPDATE
+    timestamp_oracle SET write_ts = GREATEST(write_ts, $write_ts),
+    read_ts = GREATEST(read_ts, write_ts)
+    WHERE timeline = $timeline;
+```
+
+> [!NOTE]
+> All of these queries go through the 1-phase commit fast path in CRDB: the
+> output for `EXPLAIN ANALYZE (VERBOSE)` contains `auto commit`.
+
+TODO: Actually re-check this for the final state of those queries.
+
+An important thing to note here is that timestamp operations are now correct by
+themselves: we don't need the `confirm_leadership()` call anymore for
+preventing linearization anomalies.
+
+## Optimizations that are required to make the distributed TimestampOracle viable
+
+As part of query processing, we perform multiple oracle operations per each
+single query so simply replacing the current in-memory TimestampOracle with the
+distributed TimestampOracle would not be viable: operations that were
+previously "free" would now take multiple milliseconds.
+
+The solution for this is that we have to batch oracle operations, both for read
+and write queries. We will examine below how this is similar to the current
+code where we perform expensive `confirm_leadership()` calls, but batch them to
+amortize cost.
+
+### Batching of `get_read_ts` operations
+
+Currently, the pipeline for read queries (peeks, in internal parlance) looks like this:
+
+1. peek comes in
+2. determine read timestamp using `get_read_ts()`
+3. render query graph and send off to compute
+4. wait for results from compute
+5. when results are ready add them to a `pending_peek_results` buffer
+6. whenever we are not busy and have pending query results: perform
+   `confirm_leadership()` and send results back to client
+
+The salient points here are that the oracle operations in step #2 are "free",
+so we are doing one oracle operation per peek, and that the
+`confirm_leadership()` operation of step #6 translates to a CRDB transaction,
+which is expensive. Batching up those pending results and doing a single
+`confirm_leadership()` operation is what makes this approach viable.
+
+We have to do a similar optimization for the distributed TimestampOracle but
+instead of batching up results for `confirm_leadership()` we need to batch up
+incoming peeks and determine a timestamp for multiple peeks using one oracle
+operation. We can do this because the distributed TimestampOracle operations
+provide linearization in and of themselves and we no longer need the
+`confirm_leadership()` operation.
+
+The modified pipeline for peeks will look like this:
+
+1. peek comes in
+2. add peek to `pending_peeks`
+3. whenever we are not busy, determine a read timestamp for all pending peeks
+   using `get_read_ts()`
```

Comment:
One note is that when we start having isolated/distributed serving processes, we'll start to lose some of the benefit of amortization, since the batching will happen per serving process. The same would have been true of `confirm_leadership` so I don't think it's a big deal, just wanted to point that out.